### PR TITLE
feat(ws): initial commit for controller

### PR DESCRIPTION
**Pull request by jiridanek**
```
cd workspaces/controller
go mod init github.com/kubeflow/notebooks/workspaces/controller
kubebuilder init \
    --plugins=go/v4 \
    --domain=kubeflow.org \
    --project-name=workspace-controller \
    --repo=github.com/kubeflow/notebooks/workspaces/controller
kubebuilder create api --version v1beta1 --kind Workspace
kubebuilder create api --version v1beta1 --kind WorkspaceKind
make manifests
```

## TODO:
* [x] think about the branch patterns for GHA triggering (add release branches?)
* ~~integration test with kind (kubeflow/kubeflow notebooks controller has that)~~
   * separate PR, I'd prefer to merge this scaffold separately from that
* ~~Translate the [current CRD spec](https://github.com/thesuperzapper/kubeflow-notebooks-v2-design/tree/main/crds)~~
   * https://github.com/kubeflow/notebooks/pull/6
   * separate PR, I'd prefer to merge this scaffold separately from that, but up to you ofc
* [x] Declare gha perms/scopes, minimum privilege, because we can

maybe something else from https://docs.google.com/document/d/1SiWLah-U07hAc47sSsoI8-NkbkLnHl1YVzxw193QUKE/edit#heading=h.naudedl8te2q

### CODE REVIEW

**Comment by thesuperzapper** [comment]
@jiridanek I have made a new branch, which we can develop in (rather than `main`), please make your PR into: https://github.com/kubeflow/notebooks/tree/notebooks-v2

File: workspaces/controller/config/default/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# Adds namespace to all resources.
+namespace: workspaces-controller-system
```
Comment (thesuperzapper) [inline workspaces/controller/config/default/kustomization.yaml:2]: Lets re-init this folder with `workspace-controller` (no `"s"`)

**Review by thesuperzapper — CHANGES_REQUESTED** [review]

File: .github/workflows/workspaces-controller_unit-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+permissions: {}
+
+on:
+  push:
+    branches: [ "main", "notebooks-v2", "v*-branch" ]
+  pull_request:
+    paths: [ "workspaces/controller/**" ]
+    branches: [ "main", "notebooks-v2", "v*-branch" ]
+  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply .github/workflows/workspaces-controller_unit-tests.yaml]: removed

File: workspaces/controller/.gitignore [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+# Kubernetes Generated files - skip generated files, except for vendored files
+!vendor/**/zz_generated.*
+
+# editor and IDE paraphernalia
+.idea
+.vscode
+*.swp
+*.swo
+*~
```
[the remark this answers is not in the document]
Comment (jiridanek) [reply workspaces/controller/.gitignore]: removed

File: workspaces/controller/config/default/kustomization.yaml (hunk 1 above)
Comment (jiridanek) [reply workspaces/controller/config/default/kustomization.yaml:2]: And use latest kubebuilder. And update go to 1.22 (latest).

File: workspaces/controller/go.mod [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	golang.org/x/net v0.19.0 // indirect
+	golang.org/x/oauth2 v0.12.0 // indirect
+	golang.org/x/sys v0.16.0 // indirect
+	golang.org/x/term v0.15.0 // indirect
+	golang.org/x/text v0.14.0 // indirect
+	golang.org/x/time v0.3.0 // indirect
+	golang.org/x/tools v0.16.1 // indirect
+	gomodules.xyz/jsonpatch/v2 v2.4.0 // indirect
+	google.golang.org/appengine v1.6.7 // indirect
+	google.golang.org/protobuf v1.31.0 // indirect
```
Comment (jiridanek) [inline workspaces/controller/go.mod:60]: there happens to be a cve against this protobuf version; I guess we can fix it with dependabot or something like that

File: workspaces/controller/config/default/kustomization.yaml (hunk 1 above)
Comment (jiridanek) [reply workspaces/controller/config/default/kustomization.yaml]: renamed

the 1.22 in go.mod is not changed when I use latest kubebuilder and go 1.22 is installed on my machine, so I won't do anything about that

**Review by thesuperzapper — APPROVED** [review]

**Pull request merged**